add --keepMaxDeltaReleases to match --keepMaxReleases

BryanAldrich opened the pull request:
adds a new option for -keepMaxDeltaReleases implemented exactly like --keepMaxReleases

Commit:
add --keepMaxDeltaReleases to match --keepMaxReleases

BryanAldrich commented:
I wanted to keep only one or two full releases, but like 10 deltas.

caesay commented:
I guess I could see the argument for why someone might want this if you're really in a pinch for storage space and your app is really big. 

Two things. First, if "--keepMaxDeltas" is not specified, the behaviour should be the same as it is today (that is, both the same number of full and delta versions are kept). 

Secondly, I don't think the code works currently. Maybe I need to look at this again with fresh eyes, but pretty sure that the first time this happens:
```
toDelete = releaseEntries.Where(x => x.Version < minVersion).ToArray();
releaseEntries = releaseEntries.Except(toDelete).ToArray();
```
Both the deltas and full versions will have been removed from `releaseEntries`, and the code you copied / pasted will not do anything.

Commit:
Fix the maxDeltaReleases when maxReleases is used.
Split the variables out to be more verbose.
Fix issue where maxReleases would keep X full releases, then delete all but X delta/full releases leaving less than X full releases.

BryanAldrich commented:
You were correct, the releaseEntries was getting filtered in the --keepMaxReleases block, thus causing the DeltaReleases to never fire.
It did show me an issue, it was counting the # of full releases, then trimming the resulting releases to the same count. If it found 5 full releases, it would leave 5 total release files be they delta or full. I've adjusted that.
If you like, we can have maxDelta set to max if they don't specify it? That would keep behavior consistent with the current expected behavior. Right?

caesay commented:
> It did show me an issue, it was counting the # of full releases, then trimming the resulting releases to the same count. 
The previous / expected logic (which I think did work) was if you specified a max releases of 10, you should be left with 10 full and 10 delta releases (so 20 total). A delta+full is considered one release, I don't really consider a delta by itself a release, and there are some areas of Velopack which will completely ignore deltas if there is no corresponding full release.

Before merging this I would need to test all of the common actions using just deltas to clean up those areas which currently expect that full/delta releases will be available in pairs

Commit:
set the default command line value for keepMaxDeltaReleases to the same as keepMaxReleases to keep the new behavior the same as the old behavior.

BryanAldrich commented:
Instead my dad's carpentry saying "Measure twice, cut once". I think I need to have "Read twice, program once".

The behavior was as you said. It counted full releases, then cut off based on the version, not the number of releases.
I just checked in a change to the Cli options to use the DefaultValueFactory to set the default at runtime if --keepMaxReleases is set, but --keepMaxDeltaReleases is not.
